File: Pandacap/Podcasts/WmaZipSplitter.cs
using NAudio.MediaFoundation;
using NAudio.Wave;
using System.IO.Compression;

namespace Pandacap.Podcasts
{
    public class WmaZipSplitter(
        IHttpClientFactory httpClientFactory)
    {
        public async Task SegmentZip(
            Uri uri,
            TimeSpan segmentLength,
            Stream outputStream,
            CancellationToken cancellationToken)
        {
            string filename = uri.Segments.Last();
            string basename = Path.GetFileNameWithoutExtension(filename);

            string tempfile = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid()}-{filename}");

            try
            {
                MediaFoundationApi.Startup();

                using (var client = httpClientFactory.CreateClient())
                {
                    using var resp = await client.GetAsync(uri, cancellationToken);
                    using var stream = await resp.Content.ReadAsStreamAsync(cancellationToken);
                    using var fs = new FileStream(tempfile, FileMode.CreateNew, FileAccess.Write);
                    await stream.CopyToAsync(fs, cancellationToken);
                }

                using var reader = new AudioFileReader(tempfile);

                byte[] buffer = new byte[
                    reader.WaveFormat.SampleRate
                    * (long)segmentLength.TotalSeconds
                    * reader.WaveFormat.Channels
                    * reader.WaveFormat.BitsPerSample
                    / 8];

                using var archive = new ZipArchive(
                    outputStream,
                    ZipArchiveMode.Create,
                    leaveOpen: false);

                int i = 0;
                while (true)
                {
                    int read = await reader.ReadAsync(buffer, cancellationToken);
                    if (read == 0)
                        break;

                    using var inputStream = new MemoryStream(buffer, 0, read, writable: false);
                    using var outputBuffer = new MemoryStream();

                    MediaFoundationEncoder.EncodeToWma(
                        new WaveProvider(
                            inputStream,
                            reader.WaveFormat),
                        outputBuffer);

                    var entry = archive.CreateEntry($"{basename}_{++i:00}.wma", CompressionLevel.NoCompression);

                    using var zipStream = entry.Open();

                    outputBuffer.Position = 0;
                    await outputBuffer.CopyToAsync(zipStream, cancellationToken);
                }
            }
            finally
            {
                if (File.Exists(tempfile))
                    File.Delete(tempfile);
            }
        }
    }
}

File: Pandacap.Podcasts/WmaZipSplitter.fs
namespace Pandacap.Podcasts

open System
open System.IO
open System.Net.Http
open System.Threading
open NAudio.MediaFoundation
open System.IO.Compression
open NAudio.Wave

type WmaZipSplitter(httpClientFactory: IHttpClientFactory) =
    let _ = MediaFoundationApi.Startup()

    let readAudioFile (segmentLength: TimeSpan) (filename: string) = seq {
        use reader = new AudioFileReader(filename)

        let length =
            reader.WaveFormat.SampleRate
            * int segmentLength.TotalSeconds
            * reader.WaveFormat.Channels
            * (reader.WaveFormat.BitsPerSample / 8)

        let mutable finished = false
        while not finished do
            let buffer = Array.zeroCreate<byte> length
            let read = reader.Read(buffer)
            if read = 0 then
                finished <- true
            else
                let stream = new MemoryStream(buffer, 0, read, writable = false)

                yield {
                    new IWaveProvider with
                        override _.Read(buffer, offset, count) = stream.Read(buffer, offset, count)
                        override _.WaveFormat = reader.WaveFormat
                }
    }

    let encodeAudioSegment (waveProvider: IWaveProvider) =
        use output = new MemoryStream()
        MediaFoundationEncoder.EncodeToWma(waveProvider, output)
        output.ToArray()

    let createZip (outputStream: Stream) (encodedSegments: byte array seq) =
        let mutable track = 1

        use archive = new ZipArchive(
            outputStream,
            ZipArchiveMode.Create,
            leaveOpen = false)

        for encodedSegment in encodedSegments do
            let entry = archive.CreateEntry(
                sprintf "%02d.wma" track,
                CompressionLevel.NoCompression)

            use zipStream = entry.Open()
            zipStream.Write(encodedSegment)

            track <- track + 1

    member _.SegmentZip(uri: Uri, segmentLength: TimeSpan, outputStream: Stream, cancellationToken: CancellationToken) = task {
        let filename = Array.last uri.Segments

        let tempFile = Path.Combine(
            Path.GetTempPath(),
            $"{Guid.NewGuid()}{Path.GetExtension(filename)}")

        try
            do! task {
                use client = httpClientFactory.CreateClient()
                use! resp = client.GetAsync(uri, cancellationToken)
                use! stream = resp.EnsureSuccessStatusCode().Content.ReadAsStreamAsync()
                use fs = new FileStream(tempFile, FileMode.CreateNew, FileAccess.Write)
                do! stream.CopyToAsync(fs, cancellationToken)
            }

            tempFile
            |> readAudioFile segmentLength
            |> Seq.map encodeAudioSegment
            |> createZip outputStream
        finally
            if File.Exists(tempFile) then
                File.Delete(tempFile)
    }
